tests/beaker_tests/README-beaker-script-ref.md

```diff
@@ -0,0 +1,232 @@
+# Beaker Test Script Guide to Test Inputs
+
+#### Table of Contents
+
+* [Overview](#overview)
+* [Test Prerequisites](#tp)
+* [Return Codes](#rc)
+* [The `tests` Hash](#tests)
+ * [Top-level Keys](#tlk)
+ * [Test Case Keys](#tck)
+* [Title Pattern Testing](#tpt)
+
+## <a name="overview">Overview</a>
+
+The beaker test scripts have gone through a number of rewrites in an effort to reduce the time required to develop new test scripts. The current revision uses a `tests` hash as a container for all of the inputs required to fully run each test case. This document is a common reference for the `tests` hash and other required test variables.
+
+## <a name="tp">Test Prerequisites</a>
+
+Each test case in each test file makes basic assumptions about the state of the test environment:
+
+* The beaker host configuration file is set up with proper access credentials for agent and master
+* The Puppet Master daemon is running
+* SSH is enabled in the agent environment
+* The agent SSL certificate has been generated and signed by the Puppet Master
+
+## <a name="rc">Return Codes</a>
+
+The beaker tests will validate tests using return codes from bash and puppet commands:
+
+#### Bash Command Return Codes
+
+* `0` = successful command execution
+* `> 0` = unsuccessful command execution
+
+#### Puppet Command Return Codes
+
+* `0` = No Changes have occurred (idempotence tests)
+* `1` = An Error has been raised
+* `2` = Changes have occurred (manifest tests)
+* `4` = A Failure has occurred
+* `6` = A Change and Failure has occurred
+
+*Note: The `on()` method returns `0` as its default exit code*
+
+## <a name="tests">The `tests` hash</a>
+
+The tests hash is a convenient container for common and custom test variables. Common variables that are used by all test cases are defined as top-level keys, while each test case is a top-level key with the test case variables defined as the key-value; for example:
+
+```
+# A top-level key for common variables
+tests = {
+  resource_name: 'cisco_bgp_neighbor_af',
+}
+
+# A test-case key with test variables within
+tests[:bgp_weight] = {
+  manifest_props: { weight: '30' },
+}
+```
+
+Custom keys may be added for various unique test requirements but they should be documented in this guide.
+
+### <a name="tlk">Top-level Keys</a>
+
+*Req = Required, Opt = Optional*
+
+|     | Key Name         | Value and Usage
+|:--- |:-----------------|:---------------
+| Req | `:agent`         | The `agent` object variable name. It is used for accessing the agent environment.
+| Req | `:master`        | The `master` object variable name. It is used for accessing the puppet master environment.
+| Req | `:resource_name` | The resource name used in the test. This is used with the `puppet resource` command as well as the testheader value in some tests.
```

Comment:
Yeah, I'd like to just get rid of it altogether. I don't use it in the new test files but some of the utilitylib methods still have it as a local variable. More:

* `testheader` is a standalone variable in the old tests
* I had a :testheader key in some of the tests to use with the skip logic but it's really just a log message and seems unnecessary
* After I created `:resource_name` I realized I could overload its use in the skip logic (`testheader = 'Resource cisco_foo'`, while `tests[:resource_name] = 'cisco_foo'` so I yanked it out of the new test files.